cli/src/lib.rs

```diff
@@ -494,21 +512,25 @@ pub fn read_config(
 ) -> Result<(Configuration, Option<GenesisNetwork>)> {
     let config = ConfigurationProxy::default();
 
-    let config = if let Some(actual_config_path) = path.try_resolve() {
+    let config = if let Some(actual_config_path) = path
+        .try_resolve()
+        .wrap_err("Failed to resolve configuration file")?
+    {
         let mut cfg = config.override_with(ConfigurationProxy::from_path(&*actual_config_path));
+        let config_dir = actual_config_path
+            .parent()
+            .expect("If config file was read, than it should has a parent. It is a bug.");
 
         // careful here: `genesis.file` might be a path relative to the config file.
         // we need to resolve it before proceeding
         // TODO: move this logic into `iroha_config`
-        if let Some(genesis) = &mut cfg.genesis {
-            if let Some(Some(genesis_path)) = &genesis.file {
-                let genesis_path = PathBuf::from(genesis_path);
-                if genesis_path.is_relative() {
-                    let resolved = actual_config_path.join(genesis_path);
-                    genesis.file = Some(Some(resolved.to_string_lossy().into_owned()));
-                }
-            }
-        }
+        let join_to_config_dir = |x: &mut PathBuf| {
+            *x = config_dir.join(&x);
+        };
+        mutate_nested_option!(cfg, [genesis, file, self], join_to_config_dir);
+        mutate_nested_option!(cfg, [snapshot, dir_path], join_to_config_dir);
+        mutate_nested_option!(cfg, [kura, block_store_path], join_to_config_dir);
+        mutate_nested_option!(cfg, [telemetry, file, self], join_to_config_dir);
```

Comment:
It wouldn't. While `snapshot.dir_path` and `kura.block_store_path` are `Option<T>`, `genesis.file` and `telemetry.file` are `Option<Option<T>>` (because they are optional in the final config).